src/Features/Core/Portable/ExtractMethod/MethodExtractor.VariableSymbol.cs

```diff
@@ -136,40 +136,32 @@ public int CompareTo(ParameterVariableSymbol other)
                 return (_parameterSymbol.Ordinal > other._parameterSymbol.Ordinal) ? 1 : -1;
             }
 
-            private int CompareTo(IMethodSymbol left, IMethodSymbol right)
+            private int CompareMethodParameters(IMethodSymbol left, IMethodSymbol right)
             {
                 if (left == null && right == null)
                 {
+                    // not method parameters
                     return 0;
                 }
 
                 if (left.Equals(right))
                 {
+                    // parameter of same method
                     return 0;
                 }
 
-                if (left.MethodKind == MethodKind.AnonymousFunction &&
-                    right.MethodKind != MethodKind.AnonymousFunction)
-                {
-                    return 1;
-                }
+                // these methods can be either regular one, anonymous function, local function and etc
+                // but all must belong to same outer regular method.
+                // so, it should have location pointing to same tree
+                var leftLocations = left.Locations;
+                var rightLocations = right.Locations;
 
-                if (left.MethodKind != MethodKind.AnonymousFunction &&
-                    right.MethodKind == MethodKind.AnonymousFunction)
-                {
-                    return -1;
-                }
-
-                if (left.MethodKind == MethodKind.AnonymousFunction &&
-                    right.MethodKind == MethodKind.AnonymousFunction)
-                {
-                    Contract.ThrowIfFalse(left.Locations.Length == 1);
-                    Contract.ThrowIfFalse(right.Locations.Length == 1);
+                var commonTree = leftLocations.Select(l => l.SourceTree).Intersect(rightLocations.Select(l => l.SourceTree)).WhereNotNull().First();
 
-                    return left.Locations[0].SourceSpan.Start - right.Locations[0].SourceSpan.Start;
-                }
+                var leftLocation = leftLocations.First(l => l.SourceTree == commonTree);
+                var rightLocation = rightLocations.First(l => l.SourceTree == commonTree);
 
-                return Contract.FailWithReturn<int>("Shouldn't reach here");
+                return leftLocation.SourceSpan.Start - rightLocation.SourceSpan.Start;
```

Comment:
here, we crash because it doesn't expect local function which added after extract method.

Reply:
if I write it future proof, I can't assume that common tree must exist here as well, since who knows what kind of new method symbol we would add in future?

Reply:
Yes.  i would expect this code to not have a Contract.  I would expect this to either be resilient to a new sort of symbol being passed in.  Or the code should be able to indicate that it doesn't know what is going on, and the refactoring should be suppressed.  

Assuming any of our symbol kinds is "complete" is not something i think is acceptable for the LS code to do.  It can NFW if we want to send us data to help fill in these gaps without user reports.  But it should not crash just because it expected a certain set of methods, and then new ones were added.

The only way this sort of code is acceptable is if the author has put a system in place to crash *immediately* the moment a new enum member is added.  i.e. that way when someone on the compilre team adds MethodKind.LocalFunction, then the LS refuses to even *load* because you have a static constructor that fails.  That way the person adding the enum member knows to go fix up this code.

Right now you're basically saying: "i've only written code to handle a fixed set of cases, and i will crash in the future if any new cases happen.  But there's no way for you to know this will happen and customers will be hit by this in the wild."   That's not responsible.  Either we need to be resilient to our enums/symbols expanding over time.  or we must fail *immediately* so that hte person adding the new language feature knows they have to fix/test this code.

Reply:
> since who knows what kind of new method symbol we would add in future?

Correct.  We need to be resilient here.  For example: look at AbstractSymbolDescriptionBuilder.AddDescriptionPartAsync.  It is needed to create a description for an arbitrary symbol.  It hardcodes is knowledge about which symbols are supported, but it has a standard 'fall through' case for any symbols it doesn't expect.  It doesn't crash saying "i didn't know how to create quick info for this new language feature type."

Reply:
like I said, fine opening issue for what you are saying, not going to make that change in this PR.  and I don't believe what you are saying is completely true since if that is true, when new language feature is added from compiler, IDE should never crash which is not true at all. 

"The only way this sort of code is acceptable is if the author has put a system in place to crash immediately" not sure what you meant immediately. you mean when there is no one invoking the feature, it need to somehow feature that out immediately?

Reply:
I really didn't want to. but you put the symbol display as example, so I dug in to see whether it is really future proof and there is no Contract.

it does have some such as this
http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AbstractSymbolDescriptionBuilder.cs,303

if new group kind is added, it will throw. so, now will it immediately throw when group kind is added and vs run first time? no it won't until quick info is shown. will it throw immediately when quick info is shown? no it won't until quickinfo that has new group kind is needed. if new group kind is needed, will it immediately throw? no it won't until multiple line is needed in description.

..

this kind of code in Roslyn is common. and this actually helped us a lot when something changed. sure we can move to new pattern where we try to not use this kind of pattern as much like we are now moving away from fatal watson to NFW.

Reply:
by the way, this doesn't crash VS. it shows info bar.

Reply:
That's not a good case.  SymbolDescriptionGroups is tightly coupled with AbstractSymbolDescriptionBuilder.  You can of course have tightly coupled code have knowledge of itself.  The issue is when you cross layers, or use subsystems that are *by design* expected to change over time.  The compiler layer is one of those systems.

Reply:
well, don't want to argue more. I see this differently so. like I said I am fine to open an issue. I just am not going to change how extract method handle unexpected condition in this PR. 

if this crash VS, I will do something not to crash VS. but it doesn't.